src/t_hash.c

```diff
@@ -260,6 +260,74 @@ int hashTypeSet(robj *o, sds field, sds value, int flags) {
     return update;
 }
 
+/* Similar to hashTypeSet(), but only add a new field if it does not exist. Otherwise, do nothing.
+ * Return 0 if the new field was added.
+ * Return 1 if the field already exists.
+ * Do not call hashTypeTryConversion() before this function. The conversion will be done in this
+ * function if needed.
+ */
+int hashTypeSetnx(robj *o, sds field, sds value, int flags) {
```

Comment:
Did some skimming instead of thorough reviewing, looks pretty nice.
Just wondering why you wrote a new function instead of doing some small refactors in `hashTypeSet()` , like adding extra flag (maybe just a new bit with `flags` para), to avoid duplicated lpFind or dictFind?
In this way may it be simpler than writing a new function I guess?

Reply:
Thanks for your review.

The `hsetnxCommand()` may `not` insert the field; in this case we should not check the conversion.

Before this change, the code looks like this:
```c
if (hashTypeExists(o, c->argv[2]->ptr)) {
         ...
} else {
        hashTypeTryConversion(o,c->argv,2,3);
        hashTypeSet(o,c->argv[2]->ptr,c->argv[3]->ptr,HASH_SET_COPY);
        ...
}
```

We can see that the `hashTypeTryConversion()` only happened after we sure know that the field does not exist.

And here is the code after this change:

```c
if (hashTypeSetnx(o, c->argv[2]->ptr, c->argv[3]->ptr, HASH_SET_COPY)) {
        ...
} else {
        ...
}
```


No matter what we call(`hashTypeSet()` or `hashTypeSetnx()`), we shouldn't call `hashTypeTryConversion()` before it because at this time we do not know whether the field exists or not. So the conversion should be done inside the function `hashTypeSet()` or `hashTypeSetnx()`. However, in most cases the function `hashTypeSet()` is called after `hashTypeTryConversion()`, which means many checks for conversion inside `hashTypeSet()` is useless. So I think this will make things a little confusing and introduce too many `if-branch` if we use `hashTypeSet()`. So I wrote a new function for this. :)